### PR TITLE
Cawllec/fix shouldnt notify logging

### PR DESCRIPTION
**Pull request by Cawllec**
Disables delivery errors in the case that shouldNotify is false, adds configuration option to overwrite this behaviour if necessary

### CODE REVIEW

File: lib/bugsnag.js [hunk 1]
```diff
@@ -92,7 +103,7 @@ Bugsnag.notify = function(error, options, cb) {
         options = {};
     }
     if (!Bugsnag.shouldNotify()) {
-        if (cb) {
+        if (cb && Configuration.logAllDeliveryErrors) {
```
Comment (bengourley) [inline lib/bugsnag.js]: I think this branch could be in a more appropriate place, as the the option mentions `log` but no logging is done here (an `Error` object is created and passed to the `cb()`).

A better place would be in the definition of [`autoNotifyCallback`](https://github.com/bugsnag/bugsnag-node/blob/37c0c7bb67d551b5c1fdf57d3e324367a4f7e3d4/lib/bugsnag.js#L20-L25).

File: lib/bugsnag.js [hunk 2]
```diff
@@ -68,6 +68,17 @@ Bugsnag.register = function(apiKey, options) {
...
 
+    // If notify release stages is set and we're not going to be notifying bugsnag we need to log
+    // this once and then prevent it from being logged in the future
+    if (!Bugsnag.shouldNotify()) {
+        Configuration.logger.warn(Configuration.apiKey ?
+            "Current release stage not permitted to send events to Bugsnag." :
+            "Bugsnag has not been configured with an api key!");
+        Configuration.logger.warn(Configuration.logAllDeliveryErrors ?
+            "Delivery errors will still be reported" :
+            "Delivery errors will not be reported. Configure logAllDeliveryErrors to change this behaviour");
```
Comment (bengourley) [inline lib/bugsnag.js]: I think these log messages are confusing and could be worded better…

Something along the lines of:
"Calls to bugsnag.notify() will silently fail"
or
"Nothing will be sent to Bugsnag. Any bugsnag.notify() calls are essentially NOOPs."

File: lib/configuration.js [hunk 3]
```diff
@@ -24,6 +24,7 @@ var Configuration = {
     metaData: {},
     logger: new Logger(),
     sendCode: true,
+    logAllDeliveryErrors: false,
```
Comment (bengourley) [inline lib/configuration.js]: The name of this option might want changing to be consistent with the log message.

**Comment by Cawllec** [comment]
Based on discussion with @bengourley I've removed the changes based around the configuration options, and instead ensured that when an error occurs and is correctly not-notified the original error is logged instead of the notification error.

This still leaves the issue with the initial log level however.

**Comment by bengourley** [comment]
This is better!

Here is how the user's [repro repo](https://github.com/keattang/bugsnag-bug-repro) now behaves:

image

I think it would be good if the error stack got into the log and looked more like how uncaught errors get logged out:

image

appending `e.stack` to the log message should do the trick?

**Comment by Cawllec** [comment]
Appended that stack as suggested and removed the `Bugsnag:` repetition.

I notice that it's logging each issue twice, any idea why?

File: lib/bugsnag.js [hunk 4]
```diff
@@ -12,13 +12,16 @@ var domain = require("domain"),
...
 function autoNotifyCallback(notifiedError, uncaughtError) {
     if (!uncaughtError) {
         uncaughtError = notifiedError.domain;
     }
     return function(error) {
         if (error) {
-            Configuration.logger.error("Bugsnag: error notifying bugsnag.com - " + error);
+            Configuration.logger.error(Bugsnag.shouldNotify() ?
+                "error notifying bugsnag.com - " + error + error.stack :
+                "Bugsnag will not be notified - " + notifiedError + notifiedError.stack);
```
Comment (bengourley) [inline lib/bugsnag.js]: `notifiedError` will be coerced into a string, which will contain the `Error: <message>` bit, but then the string `notifiedError.stack` also contains `Error: <message>` followed by each stack line.

In summary you only need `.stack` and not both. This goes for line 23 and 24.

I've had a tinker and coming at it from the user's perspective, I think the best output would be something like this…

```
Bugsnag:  configuration prevents the following error from being sent to Bugsnag.
https://docs.bugsnag.com/platforms/nodejs/other/configuration-options/#notifyreleasestages

Error: Some error occured
    at app.use (/Users/bengourley/Development/bugsnag-bug-repro/index.js:13:11)
    at dispatch (/Users/bengourley/Development/bugsnag-bug-repro/node_modules/koa-compose/index.js:42:32)
    at /Users/bengourley/Development/bugsnag-bug-repro/node_modules/koa-compose/index.js:34:12
    at Server.handleRequest (/Users/bengourley/Development/bugsnag-bug-repro/node_modules/koa/lib/application.js:136:14)
    at emitTwo (events.js:125:13)
    at Server.emit (events.js:213:7)
    at parserOnIncoming (_http_server.js:602:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:116:23)
```

Which you should be able to do with something like…

```js
[
  "configuration prevents the following error from being sent to Bugsnag.",
  "https://docs.bugsnag.com/platforms/nodejs/other/configuration-options/#notifyreleasestages",
  "",
  err.stack
].join('\n')
```

**Comment by bengourley** [comment]
LGTM 👍

**Pull request merged**